src/Conversion/ONNXToKrnl/Tensor/ArgMax.cpp

```diff
@@ -0,0 +1,170 @@
+//===---------------- ArgMax.cpp - Lowering ArgMax Op -------------------===//
+//
+// This file lowers the ONNX ArgMax Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/ONNX/ONNXShapeHelper.hpp"
+
+using namespace mlir;
+
+struct ONNXArgMaxOpLowering : public ConversionPattern {
+  ONNXArgMaxOpLowering(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXArgMaxOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    // Gather info.
+    auto loc = op->getLoc();
+    MemRefType reducedMemRefType =
+        convertToMemRefType(*op->result_type_begin());
+    int64_t reducedRank = reducedMemRefType.getRank();
+
+    ONNXArgMaxOpAdaptor operandAdaptor(operands);
+    Value data = operandAdaptor.data();
+    auto dataType = data.getType().cast<MemRefType>();
+    int64_t dataRank = dataType.getRank();
+
+    ONNXArgMaxOp argMaxOp = llvm::cast<ONNXArgMaxOp>(op);
+    int64_t axis = argMaxOp.axis();
+    assert(axis >= -dataRank && axis <= dataRank - 1);
+    axis = axis >= 0 ? axis : (dataRank + axis);
+
+    int64_t keepdims = argMaxOp.keepdims();
+    bool isKeepdims = (keepdims == 1) ? true : false;
+
+    // Get type information
+    auto reducedMemrefShape = reducedMemRefType.getShape();
+    auto reducedElementType = reducedMemRefType.getElementType();
+    llvm::SmallVector<int64_t, 1> axes;
+    axes.push_back(axis);
+    std::map<int64_t, int64_t> outInDimMap =
+        getReductionMapping(dataType, llvm::makeArrayRef(axes), isKeepdims);
+
+    // Insert alloc and dealloc
+    bool insertDealloc = checkInsertDealloc(op);
+    Value alloc;
+    if (hasAllConstantDimensions(reducedMemRefType))
+      alloc = insertAllocAndDealloc(
+          reducedMemRefType, loc, rewriter, insertDealloc);
+    else {
+      SmallVector<Value, 2> allocOperands;
+      for (decltype(reducedRank) i = 0; i < reducedRank; ++i) {
+        if (reducedMemrefShape[i] < 0) {
+          auto dim = rewriter.create<DimOp>(loc, data, outInDimMap[i]);
+          allocOperands.push_back(dim);
+        }
+      }
+      alloc = rewriter.create<AllocOp>(loc, reducedMemRefType, allocOperands);
+      if (insertDealloc) {
+        auto *parentBlock = alloc.getDefiningOp()->getBlock();
+        auto dealloc = rewriter.create<DeallocOp>(loc, alloc);
+        dealloc.getOperation()->moveBefore(&parentBlock->back());
+      }
+    }
+
+    // There are two required and one optional Krnl loops:
+    // - One to initialize the result memref,
+    // - One to do reduction, and
+
+    // 1. Define loops to initialize the result.
+    std::vector<Value> originalLoopsInit;
+    defineLoops(rewriter, loc, originalLoopsInit, reducedRank);
```

Comment:
It seems that you are using the old interface for krnl loops. The new interface will construct the Krnl loop nest as an object and operate on it, such as push bound and get induction variable. Which document did you use for Krnl dialect?

Reply:
I think you mean class `BuildKrnlLoop`. I don't know there is a Krnl document. I read code `ONNXFlattenOpLowering` and `ONNXReductionOpLowering` and do similar process for `ONNXArgMaxOpLowering`. BTW, it seems they are using the old interface.

I can change to use `BuildKrnlLoop`. If there is a guide document for krnl dialect, please let me know. Thanks!

Reply:
Yes, it is BuildKrnlLoop. 
We will discuss how to provide the document and get back to you.

Reply:
1. We think it is necessary to use the BuildKrnlLoop interface for the newly added code.
2. Sorry for that there is no proper document yet
3. You can follow the examples for other operations (such as Tile). Notice that the newly added code uses another utility, called IndexExpr. It is not absolutely needed in your case, but may be helpful in simplifying the code for mermen allocation/loop bound,  and the same logic can be used for shape inference too.
Any question, let me know.